File: pacextractor.c
#include <stddef.h> // size_t
#include <stdio.h> // fclose, fopen, fprintf, fread, fseek, ftell, fwrite, printf, remove, SEEK_END, SEEK_SET, FILE, stderr
#include <stdlib.h> // exit, free, malloc, EXIT_FAILURE, EXIT_SUCCESS
#include <string.h> // strcmp

#include "crc16.h"
#include "pacextractor.h"

_Bool debug = 0;
char namebuf[NAME_LENGTH];

char* getString(uint16_t* baseString) {
    int length = 0;
    while (baseString > 0 && *baseString != 0 && ++length < NAME_LENGTH) {
        namebuf[length-1] = *baseString & 0xff;
        baseString++;
    }
    namebuf[length] = '\0';
    return namebuf;
}

void checkCRC16(PacHeader* pacHeader, FILE* filestream);
void printPartHeaderInfo(PartitionHeader* partitionHeader);
void printPacHeaderInfo(PacHeader* pacHeader);

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: pacextractor [-d] [-c] firmware.pac\n");
        fprintf(stderr, "\t -d\t enable debug output\n");
        fprintf(stderr, "\t -c\t compute and verify CRC16\n");
        exit(EXIT_FAILURE);
    }
    char* file = argv[argc - 1];
    _Bool crc16check = 0;

    if ((argc > 2) && (strcmp(argv[1], "-d") == 0 || strcmp(argv[2], "-d") == 0))
        debug = 1;
    if ((argc > 2) && (strcmp(argv[2], "-c") == 0 || strcmp(argv[1], "-c") == 0))
        crc16check = 1;

    FILE* filestream = fopen(file, "rb");
    if (filestream == NULL) {
        fprintf(stderr, "cannot open file %s.\n", file);
        exit(EXIT_FAILURE);
    }

    fseek(filestream, 0, SEEK_END);
    int firmwareSize = ftell(filestream);
    if (firmwareSize < sizeof(PacHeader)) {
        fprintf(stderr, "file %s is not a PAC firmware\n", file);
        exit(EXIT_FAILURE);
    }

    PacHeader pacHeader;
    fseek(filestream, 0, SEEK_SET);
    size_t rb = fread(&pacHeader, 1, sizeof(PacHeader), filestream);
    if (rb <= 0) {
        fprintf(stderr, "Error while parsing PAC header\n");
        exit(EXIT_FAILURE);
    }
    if (pacHeader.dwSize != firmwareSize) {
        fprintf(stderr, "Size mismatch error. PAC may be damaged.\n");
        exit(EXIT_FAILURE);
    } else if (strcmp(getString(pacHeader.szVersion), "BP_R1.0.0") != 0) {
        fprintf(stderr, "Unsupported PAC version\n");
        exit(EXIT_FAILURE);
    }
    if (debug) printPacHeaderInfo(&pacHeader);

    if (crc16check) checkCRC16(&pacHeader, filestream);

    PartitionHeader** partHeaders = malloc(sizeof(PartitionHeader**) * pacHeader.partitionCount);
    int i;
    fseek(filestream, pacHeader.partitionsListStart, SEEK_SET);
    for (i = 0; i < pacHeader.partitionCount; i++) {
        size_t length = SIZE_OF_PARTITION_HEADER;
        partHeaders[i] = malloc(length);
        rb = fread(partHeaders[i], 1, length, filestream);
        if (rb <= 0) {
            fprintf(stderr, "Partition header error\n");
            exit(EXIT_FAILURE);
        }
        if (partHeaders[i]->length != SIZE_OF_PARTITION_HEADER) {
            fprintf(stderr, "Unknown Partition Header format found\n");
            exit(EXIT_FAILURE);
        }
        if (debug) printPartHeaderInfo(partHeaders[i]);
    }

    printf("\nExtracting...\n\n");
    char fiveSpaces[] = "    ";
    char filebuf[4096];
    for (i = 0; i < pacHeader.partitionCount; i++) {
        if (partHeaders[i]->partitionSize == 0) {
            free(partHeaders[i]);
            continue;
        }
        fseek(filestream, partHeaders[i]->partitionAddrInPac, SEEK_SET);
        char* partFile = getString(partHeaders[i]->fileName);
        remove(partFile);
        FILE* newstream = fopen(partFile, "wb");
        printf("%s%s", fiveSpaces, partFile);
        uint32_t dataSizeLeft = partHeaders[i]->partitionSize;
        while (dataSizeLeft > 0) {
            uint32_t copyLength = (dataSizeLeft > 4096) ? 4096 : dataSizeLeft;
            rb = fread(filebuf, 1, copyLength, filestream);
            if (rb != copyLength) {
                fprintf(stderr, "\nPartition image extraction error\n");
                exit(EXIT_FAILURE);
            }
            dataSizeLeft -= copyLength;
            rb = fwrite(filebuf, 1, copyLength, newstream);
            if (rb != copyLength) {
                fprintf(stderr, "\nPartition image extraction error\n");
                exit(EXIT_FAILURE);
            }
            printf("\r%d%%", 100 - ((100 * dataSizeLeft) / partHeaders[i]->partitionSize));
        }
        printf("\r%s%s\n", partFile, fiveSpaces);
        fclose(newstream);
        free(partHeaders[i]);
    }
    printf("\nDone...\n");
    free(partHeaders);
    fclose(filestream);

    return EXIT_SUCCESS;
}

void checkCRC16(PacHeader* pacHeader, FILE* filestream) {
    size_t rb;
    if (pacHeader->dwMagic == (uint32_t) PAC_MAGIC) {
        printf("Checking CRC Part 1\n");
        fseek(filestream, 0, SEEK_SET);
        unsigned char buffer[SIZE_OF_PAC_HEADER - 4];
        rb = fread(buffer, 1, SIZE_OF_PAC_HEADER - 4, filestream);
        if (rb != (SIZE_OF_PAC_HEADER - 4)) {
            fprintf(stderr, "Partition header error\n");
            exit(EXIT_FAILURE);
        }
        uint16_t crc1 = crc16(0, buffer, SIZE_OF_PAC_HEADER - 4);
        uint16_t crc1InPac = pacHeader->wCRC1;
        if (debug) printf("Computed CRC1 = %d, CRC1 in PAC = %d\n", crc1, crc1InPac);
        if (crc1 != crc1InPac) {
            fprintf(stderr, "CRC Check failed for CRC1\n");
            exit(EXIT_FAILURE);
        }
    }

    printf("Checking CRC Part 2\n");
    int BUF_SIZE = 64 * 1024;
    unsigned char buffer[BUF_SIZE];
    fseek(filestream, SIZE_OF_PAC_HEADER, SEEK_SET);
    uint16_t crc2 = 0;
    uint32_t dataSizeLeft = pacHeader->dwSize - SIZE_OF_PAC_HEADER;
    while (dataSizeLeft > 0) {
        uint32_t copyLength = (dataSizeLeft > BUF_SIZE) ? BUF_SIZE : dataSizeLeft;
        rb = fread(buffer, 1, copyLength, filestream);
        if (rb != copyLength) {
            fprintf(stderr, "\nPartition image extraction error\n");
            exit(EXIT_FAILURE);
        }
        dataSizeLeft -= copyLength;
        crc2 = crc16(crc2, buffer, copyLength);
    }
    uint16_t crc2InPac = pacHeader->wCRC2;
    if (debug) printf("Computed CRC2 = %d, CRC2 in PAC = %d\n\n", crc2, crc2InPac);
    if (crc2 != crc2InPac) {
        fprintf(stderr, "CRC Check failed for CRC2\n");
        exit(EXIT_FAILURE);
    }
}

void printPacHeaderInfo(PacHeader* pacHeader) {
    printf("Version"    "\t\t= %s"  "\n", getString(pacHeader->szVersion));
    printf("Size"       "\t\t= %u"  "\n", pacHeader->dwSize);
    printf("PrdName"    "\t\t= %s"  "\n", getString(pacHeader->productName));
    printf("FirmwareName" "\t= %s"  "\n", getString(pacHeader->firmwareName));
    printf("FileCount"    "\t= %d"  "\n", pacHeader->partitionCount);
    printf("FileOffset"   "\t= %d"  "\n", pacHeader->partitionsListStart);
    printf("Mode"       "\t\t= %d"  "\n", pacHeader->dwMode);
    printf("FlashType"    "\t= %d"  "\n", pacHeader->dwFlashType);
    printf("NandStrategy" "\t= %d"  "\n", pacHeader->dwNandStrategy);
    printf("IsNvBackup"   "\t= %d"  "\n", pacHeader->dwIsNvBackup);
    printf("NandPageType" "\t= %d"  "\n", pacHeader->dwNandPageType);
    printf("PrdAlias"     "\t= %s"  "\n", getString(pacHeader->szPrdAlias));
    printf("OmaDmPrdFlag" "\t= %d"  "\n", pacHeader->dwOmaDmProductFlag);
    printf("IsOmaDM"    "\t\t= %d"  "\n", pacHeader->dwIsOmaDM);
    printf("IsPreload"    "\t= %d"  "\n", pacHeader->dwIsPreload);
    printf("Magic"     "\t\t= %#x"  "\n", pacHeader->dwMagic);
    printf("CRC1"       "\t\t= %u"  "\n", pacHeader->wCRC1);
    printf("CRC2"       "\t\t= %u"  "\n", pacHeader->wCRC2);
    printf("\n\n");
}

void printPartHeaderInfo(PartitionHeader* partitionHeader) {
    printf("Size"      "\t\t= %d" "\n", partitionHeader->length);
    printf("FileID"    "\t\t= %s" "\n", getString(partitionHeader->partitionName));
    printf("FileName"    "\t= %s" "\n", getString(partitionHeader->fileName));
    printf("FileSize"    "\t= %u" "\n", partitionHeader->partitionSize);
    printf("FileFlag"    "\t= %d" "\n", partitionHeader->nFileFlag);
    printf("CheckFlag"   "\t= %d" "\n", partitionHeader->nCheckFlag);
    printf("DataOffset"  "\t= %u" "\n", partitionHeader->partitionAddrInPac);
    printf("CanOmitFlag" "\t= %d" "\n", partitionHeader->dwCanOmitFlag);
    printf("\n");
}

File: java/com/sprd/pacextractor/PacExtractor.java
/*
 * This program is used for unpacking .pac file of Spreadtrum Firmware used in SPD Flash Tool for flashing.
 *
 * Created : 26th January 2020
 * Author  : HemanthJabalpuri
 *
 * This file has been put into the public domain.
 * You can do whatever you want with this file.
 */

package com.sprd.pacextractor;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;

public class PacExtractor {

    public static final short SIZE_OF_PAC_HEADER = 2124; // bytes
    public static final short SIZE_OF_PARTITION_HEADER = 2580; // bytes
    public static final long PAC_MAGIC = 0xfffafffaL;

    private static RandomAccessFile in;
    private static boolean debug;

    public static void exit(String errorMsg) {
        throw new RuntimeException(errorMsg);
    }

    public static String getString(char[] string) {
        return new String(string).trim();
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: PacExtractor [-d] [-c] firmware.pac outdir");
            System.err.println("\t -d\t enable debug output");
            System.err.println("\t -c\t compute and verify CRC16");
            System.exit(2);
        }
        boolean checkCRC16 = false;
        if (args[0].equals("-d") || args[1].equals("-d")) debug = true;
        if (args[1].equals("-c") || args[0].equals("-c")) checkCRC16 = true;

        try {
            File inFile = new File(args[args.length - 2]);
            if (inFile.length() < SIZE_OF_PAC_HEADER)
                exit(inFile + " is not a PAC firmware.");

            File outdir = new File(args[args.length - 1]);
            if (!outdir.exists()) outdir.mkdirs();
            if (outdir.isFile())
                exit("File with name " + outdir + " exists.");
            String outdirS = outdir.getAbsolutePath() + File.separator;

            in = new RandomAccessFile(inFile, "r");
            PacHeader pacHeader = new PacHeader();
            if (unsigned(pacHeader.dwSize) != inFile.length())
                exit("Bin packet's size is not correct");
            else if (!getString(pacHeader.szVersion).equals("BP_R1.0.0"))
                exit("Unsupported PAC version");

            if (debug) pacHeader.printInfo();

            if (checkCRC16) checkCRC16(pacHeader);

            in.seek(unsigned(pacHeader.partitionsListStart));
            PartitionHeader[] partHeaders = new PartitionHeader[pacHeader.partitionCount];
            int i;
            for (i = 0; i < partHeaders.length; i++) {
                partHeaders[i] = new PartitionHeader();
                if (partHeaders[i].length != SIZE_OF_PARTITION_HEADER)
                    exit("Unknown Partition Header format found");
                if (debug) partHeaders[i].printInfo();
            }

            System.out.println("\nExtracting...\n");
            String fiveSpaces = "     ";
            byte[] buffer = new byte[4096];
            for (i = 0; i < partHeaders.length; i++) {
                if (partHeaders[i].partitionSize == 0)
                    continue;
                in.seek(unsigned(partHeaders[i].partitionAddrInPac));
                String outfile = getString(partHeaders[i].fileName);
                System.out.print(fiveSpaces + outfile);

                FileOutputStream fos = new FileOutputStream(outdirS + outfile);
                int read;
                long len = unsigned(partHeaders[i].partitionSize);
                long tsize = len;
                while ((read = in.read(buffer, 0, len < buffer.length ? (int) len : buffer.length)) > 0) {
                    fos.write(buffer, 0, read);
                    len -= read;
                    System.out.print("\r" + (100 - ((100 * len) / tsize)) + "%");
                }
                System.out.println("\r" + outfile + fiveSpaces);
                fos.close();
            }

            System.out.println("\nDone...");
        } catch (Exception e) {
            System.err.println(e.getMessage());
        } finally {
            try {
                if (in != null) in.close();
            } catch (IOException e) {
                System.err.println(e.getMessage());
                System.exit(1);
            }
        }
    }

    public static void checkCRC16(PacHeader pacHeader) throws IOException {
        CRC16 crc16 = new CRC16();
        int read;
        if (unsigned(pacHeader.dwMagic) == PAC_MAGIC) {
            System.out.println("Checking CRC Part 1");
            in.seek(0);
            byte[] buffer = new byte[SIZE_OF_PAC_HEADER - 4];
            read = in.read(buffer);
            for (int i = 0; i < read; ++i)
                crc16.update(buffer[i]);
            int crc1InPac = unsigned(pacHeader.wCRC1);
            if (debug) System.out.println("Computed CRC1 = " + crc16.getValue() + ", CRC1 in PAC = " + crc1InPac);
            if (crc16.getValue() != crc1InPac)
                exit("CRC Check failed for CRC1\n");
        }

        System.out.println("Checking CRC Part 2");
        crc16.reset();
        int crc2InPac = unsigned(pacHeader.wCRC2);
        byte[] buffer2 = new byte[64 * 1024];
        in.seek(SIZE_OF_PAC_HEADER);
        while ((read = in.read(buffer2)) > 0)
            for (int i = 0; i < read; i++)
                crc16.update(buffer2[i]);

        if (debug) System.out.println("Computed CRC2 = " + crc16.getValue() + ", CRC2 in PAC = " + crc2InPac + "\n");
        if (crc16.getValue() != crc2InPac)
            exit("CRC Check failed for CRC2");
    }

    public static int unsigned(short n) {
        return n & 0xffff;
    }

    public static long unsigned(int n) {
        return n & 0xffffffffL;
    }

    public static int readInt() throws IOException {
        int result = 0;
        for (int i = 0; i < 4; i++)
            result |= (in.readUnsignedByte() << (8 * i));
        return result;
    }

    public static void readInt(int[] var) throws IOException {
        for (int i = 0; i < var.length; i++)
            var[i] = readInt();
    }

    public static short readShort() throws IOException {
        short result = 0;
        for (int i = 0; i < 2; i++)
            result |= (in.readUnsignedByte() << (8 * i));
        return result;
    }

    public static void readShort(short[] var) throws IOException {
        for (int i = 0; i < var.length; i++)
            var[i] = readShort();
    }

    public static void readString(char[] var) throws IOException {
        for (int i = 0; i < var.length; i++)
            var[i] = (char) readShort();
    }

    public static class PacHeader {
        char[] szVersion = new char[24];     // packet struct version
        int dwSize;                          // the whole packet size
        char[] productName = new char[256];  // product name
        char[] firmwareName = new char[256]; // product version
        int partitionCount;                  // the number of files that will be downloaded, the file may be an operation
        int partitionsListStart;             // the offset from the packet file header to the array of PartitionHeaders start
        int dwMode;
        int dwFlashType;
        int dwNandStrategy;
        int dwIsNvBackup;
        int dwNandPageType;
        char[] szPrdAlias = new char[100];   // product alias
        int dwOmaDmProductFlag;
        int dwIsOmaDM;
        int dwIsPreload;
        int[] dwReserved = new int[200];
        int dwMagic;
        short wCRC1;
        short wCRC2;

        public PacHeader() throws IOException {
            readString(szVersion);
            dwSize = readInt();
            readString(productName);
            readString(firmwareName);
            partitionCount = readInt();
            partitionsListStart = readInt();
            dwMode = readInt();
            dwFlashType = readInt();
            dwNandStrategy = readInt();
            dwIsNvBackup = readInt();
            dwNandPageType = readInt();
            readString(szPrdAlias);
            dwOmaDmProductFlag = readInt();
            dwIsPreload = readInt();
            dwIsOmaDM = readInt();
            readInt(dwReserved);
            dwMagic = readInt();
            wCRC1 = readShort();
            wCRC2 = readShort();
        }

        public void printInfo() {
            System.out.println(
              "Version"      + "\t\t= "   + getString(szVersion)                + "\n" +
              "Size"         + "\t\t= "   + unsigned(dwSize)                    + "\n" +
              "PrdName"      + "\t\t= "   + getString(productName)              + "\n" +
              "FirmwareName" +   "\t= "   + getString(firmwareName)             + "\n" +
              "FileCount"    +   "\t= "   + unsigned(partitionCount)            + "\n" +
              "FileOffset"   +   "\t= "   + unsigned(partitionsListStart)       + "\n" +
              "Mode"         + "\t\t= "   + dwMode                              + "\n" +
              "FlashType"    +   "\t= "   + dwFlashType                         + "\n" +
              "NandStrategy" +   "\t= "   + dwNandStrategy                      + "\n" +
              "IsNvBackup"   +   "\t= "   + dwIsNvBackup                        + "\n" +
              "NandPageType" +   "\t= "   + dwNandPageType                      + "\n" +
              "PrdAlias"     +   "\t= "   + getString(szPrdAlias)               + "\n" +
              "OmaDmPrdFlag" +   "\t= "   + dwOmaDmProductFlag                  + "\n" +
              "IsOmaDM"      + "\t\t= "   + dwIsOmaDM                           + "\n" +
              "IsPreload"    +   "\t= "   + dwIsPreload                         + "\n" +
              "Magic"        + "\t\t= 0x" + Long.toHexString(unsigned(dwMagic)) + "\n" +
              "CRC1"         + "\t\t= "   + unsigned(wCRC1)                     + "\n" +
              "CRC2"         + "\t\t= "   + unsigned(wCRC2)                     + "\n\n"
            );
        }
    }

    public static class PartitionHeader {
        int length;                           // size of this struct itself
        char[] partitionName = new char[256]; // file ID,such as FDL,Fdl2,NV and etc.
        char[] fileName = new char[256];      // file name in the packet bin file. It only stores file name
        char[] szFileName = new char[256];    // Reserved now
        int partitionSize;                    // file size
        int nFileFlag;                        // if "0", means that it need not a file, and
                                              // it is only an operation or a list of operations, such as file ID is "FLASH"
                                              // if "1", means that it need a file
        int nCheckFlag;                       // if "1", this file must be downloaded;
                                              // if "0", this file can not be downloaded;
        int partitionAddrInPac;               // the offset from the packet file header to this file data
        int dwCanOmitFlag;                    // if "1", this file can not be downloaded and not check it as "All files"
                                              // in download and spupgrade tool.
        int dwAddrNum;
        int[] dwAddr = new int[5];
        int[] dwReserved = new int[249];      // Reserved for future, not used now

        public PartitionHeader() throws IOException {
            length = readInt();
            readString(partitionName);
            readString(fileName);
            readString(szFileName);
            partitionSize = readInt();
            nFileFlag = readInt();
            nCheckFlag = readInt();
            partitionAddrInPac = readInt();
            dwCanOmitFlag = readInt();
            dwAddrNum = readInt();
            readInt(dwAddr);
            readInt(dwReserved);
        }

        public void printInfo() {
            System.out.println(
              "Size"        + "\t\t= " + unsigned(length)             + "\n" +
              "FileID"      + "\t\t= " + getString(partitionName)     + "\n" +
              "FileName"    +   "\t= " + getString(fileName)          + "\n" +
              "FileSize"    +   "\t= " + unsigned(partitionSize)      + "\n" +
              "FileFlag"    +   "\t= " + nFileFlag                    + "\n" +
              "CheckFlag"   +   "\t= " + nCheckFlag                   + "\n" +
              "DataOffset"  +   "\t= " + unsigned(partitionAddrInPac) + "\n" +
              "CanOmitFlag" +   "\t= " + dwCanOmitFlag                + "\n"
            );
        }
    }
}
